custom_components/deltasol/deltasolapi.py

```diff
@@ -43,40 +44,78 @@ def __parse_data(self, response):
         return data
 
 
+    def detect_mode(self):
+        if self.api_mode is not None:
+            return self.api_mode
+        
+        try:
+            dlxurl = f"http://{self.host}/dlx/"
+            _LOGGER.debug(f"Testing DLX {dlxurl}")
+            response = requests.request("GET", dlxurl)
+            _LOGGER.debug(f"DLX response {response}")
+            _LOGGER.debug(f"Detected DLX")
+            self.api_mode = 'dlx'
+            return self.api_mode
+
+        except ValueError:
```

Comment:
The valueerror is thrown if request.json() is not able to parse the result to a json. In your case here I am pretty sure that the request gets a valid (non json answer) and KM2 device will now be set to DLX mode. So we should turn it around. I need to check what kind of answer the KM2 gives if I call the DLX way.

Reply:
You are right, I should not have left ` except ValueError:` would be better to check the status like this `if(response.status_code == 200):`
My intention was to find a way to isolate the device detection (KM2 or DLX) from a permissions check. The page of `http://host/dlx` will return a 200 on a DL2 or DL3 but I assume will give a 404 on a KM2? I have a DL2 (not in use) and a DL3 to test with, but I don't have a KM2.
Is there a similar landing page for a KM2?

Reply:
http://host/dlx will return 404 on km2
http://host/cgi-bin/resol-webservice will return 200 on km2

so yes, just implement the auto detect based on that.

and to be honest: I would stick on auto detect and dont provide a manual way for configuration. it makes it way easier for the users out there if we can auto detect it and there is no confusion about manual configuration.

Reply:
Interestingly `http://host/cgi-bin/resol-webservice` also works on the DL3. I assume it also works on the DL2. Perhaps this is the better approach? We could grab the product and use that for any conditional logic.
Can you confirm that the KM2 has a product value of KM2?
```
vendor = "RESOL"
product = "DL3"
serial = "00..."
version = "2.2.5"
build = "202102180706"
name = "DL3-00..."
features = "vbus,dl2,dl3,jobparam"
```

Reply:
that wont work cause the KM2 answers this:

`{
    "error": {
        "category": "App",
        "message": "No path info given"
    }
}`

Reply:
that is a shame.
how about this, its not as clean as I might have liked.
Still its a single call and does not need auth.
```
if(http://host/cgi-bin/resol-webservice = 200) 
  If(contains response product = "DL3" or product = "DL2") use the product.
  Else we assume KM2
```
Need to confirm that DL2 works this way, mine is in a box somewhere.

Reply:
This is the approach using a single API call `resol-webservice`.
This would be my preferred solution.
``` python
    def detect_product(self):
        if self.product is not None:
            return self.product

        try:
            url = f"http://{self.host}/cgi-bin/resol-webservice"
            _LOGGER.debug(f"Detecting resol product {url}")
            response = requests.request("GET", url)
            if(response.status_code == 200):
                _LOGGER.debug(f"response {response.text}")
                matches = re.search(r'product\s=\s["](.*?)["]', response.text)
                if matches:
                    self.product = matches.group(1).lower()
                else:
                     self.product = "km2"
                _LOGGER.debug(f"Detected {self.product}")
                return self.product
                
        except Exception as e:
            _LOGGER.debug(f"Error detecting resol product - {e}")

        _LOGGER.warning("Unable to detect resol system")
        return self.product
```

This is what I came up with for a call to `http://{self.host}s/dlx/"` and then a call to `http://{self.host}/cgi-bin/resol-webservice`
What bothers me about this approach is we are making a call the the KM2 that we know will result in a 404. Also it breaks is someone changes the order of the calls.
``` python
    def detect_mode(self):
        if self.api_mode is not None:
            return self.api_mode

        try:
            dlxurl = f"http://{self.host}s/dlx/"
            _LOGGER.debug(f"Testing DLX {dlxurl}")
            response = requests.request("GET", dlxurl)
        except Exception as e:
            _LOGGER.debug(f"Error detecting DLX - {e}")
        else:
            _LOGGER.debug(f"DLX response {response}")
            if(response.status_code == 200):
                _LOGGER.debug(f"Detected DLX")
                self.api_mode = 'dlx'
                return self.api_mode

        try:
            km2url = f"http://{self.host}/cgi-bin/resol-webservice"
            _LOGGER.debug(f"Testing KM2 {km2url}")
            response = requests.request("GET", km2url)
        except Exception as e:
            _LOGGER.debug(f"Error detecting KM2 - {e}")
        else:
            _LOGGER.debug(f"KM2 response {response}")
            if(response.status_code == 200):
                _LOGGER.debug(f"Detected KM2")
                self.api_mode = 'km2'
                return self.api_mode

        _LOGGER.warning("Unable to detect resol system")
        return self.api_mode
```

Reply:
Pushed new update, feedback welcome.

Reply:
but why don't we just do this:

```
        if self.product is not None:
            return self.product

        url = f"http://{self.host}/dlx"
        _LOGGER.debug(f"Detecting resol product {url}")
        response = requests.request("GET", url)
        if(response.status_code == 200):
            self.product = "dlx"
            _LOGGER.debug(f"Detected {self.product}")
        elif(response.status_code == 404):
            self.product = "km2"
            _LOGGER.debug(f"Detected {self.product}")
        else:
             _LOGGER.warning("Unable to detect resol system")

       return self.product
```

Reply:
this is what the KM2 gives on http://host:

```html
<!DOCTYPE html>
<html>

<head>
	<meta charset="UTF-8" />

	<meta name="viewport" content="width=device-width, initial-scale=1">
	<meta name="apple-mobile-web-app-capable" content="yes">
	<meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">

	<title>Loading...</title>

	<link rel="stylesheet" type="text/css" href="css/bootstrap.min.css" />
	<link rel="stylesheet" type="text/css" href="css/jquery-ui.min.css" />
	<link rel="stylesheet" type="text/css" href="css/ladda.css" />

	<link rel="stylesheet" media="all" href="css/base.css" />
	<link rel="stylesheet" media="only screen and (max-width: 767px)" href="css/mobile.css" />

	<link rel="shortcut icon" href="images/favicon.ico" />
</head>

<body>
	<div id="loading-overlay" class="modal-loading-spinner"></div>
	<div id="app-container"></div>

	<!-- bower:js -->
	<!-- <script src="bower_components/jquery/dist/jquery.js"></script>
    <script src="bower_components/i18next/i18next.js"></script> -->
	<!-- endbower -->

	<script src="js/bundle.js"></script>
</body>

</html>
```

Reply:
> I'd explicitly also test for a km2 somehow. So if someone inputs a wrong IP (with active webserver) it does not automatically detect a km2 but throws an unknown product error.

Agreed, would be good to explicitly test if we could without auth. Its a shame that it does not return its product in the `resol-webservice` call. The URL does have **resol** in it, so it think it's unlikely we would get a false positive on the wrong IP. Most likely we would get a 404 , host not found or some other error.

> but why don't we just do this:
` if http://host/dlx == 200 then DLX else KM2` roughly speaking :)

This means for any host for the wrong IP would result KM2.
The single call to the `resol-webservice` seems more resilient. Firstly it confirms its a resol device just by the url. For the DLX we can confirm the product, otherwise we can assume its the KM2 (same as the as the `elseif http://host/cgi-bin/resol-webservice == 200 then mode = km2`).

Reply:
You can try to use something out of the html from above. Maybe check for the title „Loading…“ and then use that as prove for KM2

Reply:
I don't have a KM2, to test that with. 
Responding to the `resol-webservice` call with a 200 is possibly better indicator that its a KM2 than seeing `Loading` on the home page. Any device that would respond to `resol-webservice` with 200 and is not a DLX device, would just as likely have the same lading home page. 

I'm not convinced another check is need, but happy to add one if we can find something that add value.

Could use something from the `resol-webservice` call response `'{ "error": { "category": "App", "message": "No path info given" } }'`, maybe confirm it contains `No path info given`, its not great but its something.

What cookies do you get back from that request? On the DL3/DL2 there is a cookie `resol-dlx-session=00FF....`. Might be something there we can use.

Reply:
there are no cookies. I created a issue at resol, maybe they know a secure way of finding it out: https://github.com/danielwippermann/resol-vbus/issues/71

Reply:
@ostat check the issue linked, the answer is there. With that we can create a brilliant auto detect mode 🥳

Reply:
@ostat don't want to make any pressure but do you have any idea if and when you would make the changes?

Reply:
Can update over the weekend :).

Reply:
Using the `GET /cgi-bin/get_resol_device_information` endpoint defiantly seems the proper way to go. This has been updated.